org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/missingelement/required_elements_missing_expectedCodeActionResolver.java

```diff
@@ -0,0 +1,89 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ * Red Hat Inc. - initial API and implementation
+ */
+package org.eclipse.lemminx.extensions.contentmodel.participants.codeactions.missingelement;
+
+import static org.eclipse.lemminx.extensions.contentmodel.participants.codeactions.missingelement.MissingElementDataConstants.DATA_ELEMENT_FIELD;
+
+import org.eclipse.lemminx.commons.BadLocationException;
+import org.eclipse.lemminx.commons.CodeActionFactory;
+import org.eclipse.lemminx.dom.DOMDocument;
+import org.eclipse.lemminx.dom.DOMElement;
+import org.eclipse.lemminx.dom.DOMNode;
+import org.eclipse.lemminx.extensions.contentmodel.model.CMDocument;
+import org.eclipse.lemminx.extensions.contentmodel.model.CMElementDeclaration;
+import org.eclipse.lemminx.extensions.contentmodel.model.ContentModelManager;
+import org.eclipse.lemminx.extensions.contentmodel.utils.XMLGenerator;
+import org.eclipse.lemminx.services.extensions.codeaction.ICodeActionResolverRequest;
+import org.eclipse.lemminx.services.extensions.codeaction.ICodeActionResolvesParticipant;
+import org.eclipse.lsp4j.CodeAction;
+import org.eclipse.lsp4j.Diagnostic;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.jsonrpc.CancelChecker;
+
+/**
+ * CodeAction resolver used to generate missing element of choice
+ * 
+ */
+
+public class required_elements_missing_expectedCodeActionResolver implements ICodeActionResolvesParticipant {
+
+	public static final String PARTICIPANT_ID = required_elements_missing_expectedCodeActionResolver.class
+			.getName();
+
+	@Override
+	public CodeAction resolveCodeAction(ICodeActionResolverRequest request, CancelChecker cancelChecker) {
+		DOMDocument document = request.getDocument();
+		Diagnostic diagnostic = request.getUnresolved().getDiagnostics().get(0);
+		CodeAction resolved = request.getUnresolved();
+
+		try {
+			Range diagnosticRange = diagnostic.getRange();
+			int startOffset = document.offsetAt(diagnosticRange.getStart()) + 1;
+			DOMNode node = document.findNodeAt(startOffset);
+
+			if (node == null || !node.isElement()) {
+				return null;
+			}
+
+			DOMElement element = (DOMElement) node;
+			Position childElementPositionStartTag = document.positionAt(element.getStartTagCloseOffset() + 1);
+			Position childElementPositionEndTag = document.positionAt(element.getEndTagOpenOffset());
+
+			Range targetRange = new Range(childElementPositionStartTag, childElementPositionEndTag);
+			XMLGenerator generator = request.getXMLGenerator(targetRange);
+
+			ContentModelManager contentModelManager = request.getComponent(ContentModelManager.class);
+
+			for (CMDocument cmDocument : contentModelManager.findCMDocument(element)) {
+				CMElementDeclaration matchesCMElement = cmDocument.findCMElement(element);
+				if (matchesCMElement != null) {
+					// Add a code action option for each element choice
+					for (CMElementDeclaration child : matchesCMElement.getElements()) {
```

Comment:
I will introduce a new method in CMElementDeclaration:

`for (CMElementDeclaration child : matchesCMElement.getPossibleRequiredElements()) {`

and create class :

```java
/*******************************************************************************
* Copyright (c) 2023 Red Hat Inc. and others.
* All rights reserved. This program and the accompanying materials
* which accompanies this distribution, and is available at
* http://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Contributors:
*     Red Hat Inc. - initial API and implementation
*******************************************************************************/
package com.thaiopensource.relaxng.pattern;

import java.util.Set;

public class PossibleRequiredElementsFunction extends AbstractPatternFunction<Void> {

	private final Set<String> requiredElementNames;

	public PossibleRequiredElementsFunction(Set<String> requiredElementNames) {
		this.requiredElementNames = requiredElementNames;
	}

	@Override
	public Void caseOther(Pattern p) {
		return null;
	}

	@Override
	public Void caseElement(ElementPattern p) {
		return caseNamed(p.getNameClass());
	}

	@Override
	public Void caseGroup(GroupPattern p) {
		return union(p);
	}

	@Override
	public Void caseChoice(ChoicePattern p) {
		// For choice, get the first pattern which is not nullable
//		Pattern p1 = p.getOperand1();
//		if (!p1.isNullable()) {
//			p1.apply(this);
//		} else {
//			Pattern p2 = p.getOperand2();
//			if (!p2.isNullable()) {
//				p2.apply(this);
//			}
//		}
//		return null;
		return union(p);
	}

	@Override
	public Void caseInterleave(InterleavePattern p) {
		return union(p);
	}

	@Override
	public Void caseAfter(AfterPattern p) {
		return p.getOperand1().apply(this);
	}

	@Override
	public Void caseOneOrMore(OneOrMorePattern p) {
		return p.getOperand().apply(this);
	}

	private Void caseNamed(NameClass nc) {
		if (!(nc instanceof SimpleNameClass))
			return null;
		requiredElementNames.add(((SimpleNameClass) nc).getName().getLocalName());
		return null;
	}
	
	private Void union(BinaryPattern p) {
		Pattern p1 = p.getOperand1();
		if (!p1.isNullable()) {
			p1.apply(this);
		}
		Pattern p2 = p.getOperand2();
		if (!p2.isNullable()) {
			p2.apply(this);
		}
		return null;
	}
}

```